File: src/main/go/leetcode/editor/cn/LruCache.go
//146.lru-cache
//运用你所掌握的数据结构，设计和实现一个 LRU (最近最少使用) 缓存机制 。
//
//
//
// 实现 LRUCache 类：
//
//
// LRUCache(int capacity) 以正整数作为容量 capacity 初始化 LRU 缓存
// int get(int key) 如果关键字 key 存在于缓存中，则返回关键字的值，否则返回 -1 。
// void put(int key, int value) 如果关键字已经存在，则变更其数据值；如果关键字不存在，则插入该组「关键字-值」。当缓存容量达到上
//限时，它应该在写入新数据之前删除最久未使用的数据值，从而为新的数据值留出空间。
//
//
//
//
//
//
// 进阶：你是否可以在 O(1) 时间复杂度内完成这两种操作？
//
//
//
// 示例：
//
//
//输入
//["LRUCache", "put", "put", "get", "put", "get", "put", "get", "get", "get"]
//[[2], [1, 1], [2, 2], [1], [3, 3], [2], [4, 4], [1], [3], [4]]
//输出
//[null, null, null, 1, null, -1, null, -1, 3, 4]
//
//解释
//LRUCache lRUCache = new LRUCache(2);
//lRUCache.put(1, 1); // 缓存是 {1=1}
//lRUCache.put(2, 2); // 缓存是 {1=1, 2=2}
//lRUCache.get(1);    // 返回 1
//lRUCache.put(3, 3); // 该操作会使得关键字 2 作废，缓存是 {1=1, 3=3}
//lRUCache.get(2);    // 返回 -1 (未找到)
//lRUCache.put(4, 4); // 该操作会使得关键字 1 作废，缓存是 {4=4, 3=3}
//lRUCache.get(1);    // 返回 -1 (未找到)
//lRUCache.get(3);    // 返回 3
//lRUCache.get(4);    // 返回 4
//
//
//
//
// 提示：
//
//
// 1 <= capacity <= 3000
// 0 <= key <= 3000
// 0 <= value <= 104
// 最多调用 3 * 104 次 get 和 put
//
// Related Topics 设计
// 👍 1438 👎 0

package main

//leetcode submit region begin(Prohibit modification and deletion)
import (
	. "container/list"
	"fmt"
)

type LRUCache struct {
	//value为节点对象
	m map[int]*Element
	l *List
	c int
}
type KV struct {
	k int
	v int
}

func Constructor(capacity int) LRUCache {
	return LRUCache{
		c: capacity,
		l: New(),
		m: make(map[int]*Element),
	}
}

func (this *LRUCache) Get(key int) int {
	var v, ok = this.m[key]
	if ok {
		//move the value to first position
		this.l.MoveToFront(v)
		return v.Value.(KV).v
	} else {
		return -1
	}
}

func (this *LRUCache) Put(key int, value int) {
	var v, ok = this.m[key]
	if ok { //key exists
		//update the value at first position
		v.Value = KV{
			k: key,
			v: value,
		}
		this.l.MoveToFront(v)
	} else { //key not exists
		//not full
		if this.l.Len() >= this.c {
			// remove the last visited from map
			delete(this.m, this.l.Back().Value.(KV).k)
			// remove the last visited from list
			this.l.Remove(this.l.Back())
		}
		//insert new key at first position
		this.l.PushFront(KV{
			k: key,
			v: value,
		})
		this.m[key] = this.l.Front()
	}

}

/**
 * Your LRUCache object will be instantiated and called as such:
 * obj := Constructor(capacity);
 * param_1 := obj.Get(key);
 * obj.Put(key,value);
 */
//leetcode submit region end(Prohibit modification and deletion)

//test from here
func main() {
	var lru = Constructor(2)
	lru.Put(1, 1)
	lru.Put(2, 2)
	fmt.Println(lru.Get(1), lru.Get(1) == 1)
	lru.Put(3, 3)
	fmt.Println(lru.Get(2), lru.Get(2) == -1)
	lru.Put(4, 4)
	fmt.Println(lru.Get(1), lru.Get(1) == -1)
	fmt.Println(lru.Get(3), lru.Get(3) == 3)
	fmt.Println(lru.Get(4), lru.Get(4), lru.Get(4) == 4)

}

File: src/main/python/leetcode/editor/en/LruCache.py
# 146.lru-cache
"""
Design a data structure that follows the constraints of a Least Recently Used (
LRU) cache. 

 Implement the LRUCache class: 

 
 LRUCache(int capacity) Initialize the LRU cache with positive size capacity. 
 int get(int key) Return the value of the key if the key exists, otherwise 
return -1. 
 void put(int key, int value) Update the value of the key if the key exists. 
Otherwise, add the key-value pair to the cache. If the number of keys exceeds the 
capacity from this operation, evict the least recently used key. 
 

 The functions get and put must each run in O(1) average time complexity. 

 
 Example 1: 

 
Input
["LRUCache", "put", "put", "get", "put", "get", "put", "get", "get", "get"]
[[2], [1, 1], [2, 2], [1], [3, 3], [2], [4, 4], [1], [3], [4]]
Output
[null, null, null, 1, null, -1, null, -1, 3, 4]

Explanation
LRUCache lRUCache = new LRUCache(2);
lRUCache.put(1, 1); // cache is {1=1}
lRUCache.put(2, 2); // cache is {1=1, 2=2}
lRUCache.get(1);    // return 1
lRUCache.put(3, 3); // LRU key was 2, evicts key 2, cache is {1=1, 3=3}
lRUCache.get(2);    // returns -1 (not found)
lRUCache.put(4, 4); // LRU key was 1, evicts key 1, cache is {4=4, 3=3}
lRUCache.get(1);    // return -1 (not found)
lRUCache.get(3);    // return 3
lRUCache.get(4);    // return 4
 

 
 Constraints: 

 
 1 <= capacity <= 3000 
 0 <= key <= 10⁴ 
 0 <= value <= 10⁵ 
 At most 2 * 10⁵ calls will be made to get and put. 
 

 Related Topics Hash Table Linked List Design Doubly-Linked List 👍 22757 👎 120
6

"""


# leetcode submit region begin(Prohibit modification and deletion)
class LRUCache:
    map: dict[int, "LRUCache.Node"]
    __slots__ = ("cap", "map", "list")

    class Node:
        pre: "LRUCache.Node"
        next: "LRUCache.Node"
        __slots__ = ("key", "val", "pre", "next")

        def __init__(self, key: int, val: int):
            self.key = key
            self.val = val
            self.pre = None
            self.next = None

        def detach(self):
            self.pre.next = self.next
            self.next.pre = self.pre

    class LinkedList:
        head: "LRUCache.Node"
        tail: "LRUCache.Node"
        __slots__ = ("head", "tail")

        def __init__(self):
            self.head = LRUCache.Node(-1, -1)
            self.tail = LRUCache.Node(-1, -1)
            self.head.next = self.tail
            self.tail.pre = self.head

        def move_to_front(self, node: "LRUCache.Node"):
            node.detach()
            self.add_to_front(node)

        def add_to_front(self, node: "LRUCache.Node"):
            first = self.head.next
            self.head.next = node
            node.pre = self.head
            node.next = first
            first.pre = node

        def remove_from_tail(self) -> "LRUCache.Node":
            last = self.tail.pre
            last.detach()
            return last

    def __init__(self, capacity: int):
        self.cap = capacity
        self.map = {}  # key -> Node
        self.list = self.LinkedList()

    def get(self, key: int) -> int:
        node = self.map.get(key)
        if node:
            self.list.move_to_front(node)
            return node.val
        return -1

    def put(self, key: int, value: int) -> None:
        if self.cap == 0:
            return
        node = self.map.get(key)
        if node:
            node.val = value
            self.list.move_to_front(node)
            return
        if len(self.map) >= self.cap:
            r = self.list.remove_from_tail()
            del self.map[r.key]
        node = LRUCache.Node(key, value)
        self.map[key] = node
        self.list.add_to_front(node)


# Your LRUCache object will be instantiated and called as such:
# obj = LRUCache(capacity)
# param_1 = obj.get(key)
# obj.put(key,value)
# leetcode submit region end(Prohibit modification and deletion)

# test from here
if __name__ == '__main__':
    lru = LRUCache(2)
    lru.put(1, 1)
    lru.put(2, 2)
    lru.put(3, 3)
    print(lru.get(1))
